mithril-stm/src/stm.rs

```diff
@@ -853,32 +838,215 @@ impl<D: Clone + Digest + FixedOutput + Send + Sync> StmAggrSig<D> {
+        let sig_reg_size = usize::try_from(u64::from_be_bytes(u64_bytes))
             .map_err(|_| StmAggregateSignatureError::SerializationError)?;
 
-        let mut signatures = Vec::with_capacity(size);
+        let mut sig_reg_list = Vec::with_capacity(size);
         for i in 0..size {
-            signatures.push((
-                StmSig::from_bytes::<D>(
-                    &bytes[16 + i * (sig_size + 104)..16 + sig_size + i * (sig_size + 104)],
-                )?,
-                MTLeaf::from_bytes(
-                    &bytes[16 + sig_size + i * (sig_size + 104)..16 + (i + 1) * (sig_size + 104)],
-                )
-                .map_err(|_| StmAggregateSignatureError::SerializationError)?,
-            ));
+            let sig_reg = StmSigRegParty::from_bytes::<D>(
+                &bytes[16 + (sig_reg_size * i)..16 + (sig_reg_size * (i + 1))],
+            )?;
+            sig_reg_list.push(sig_reg);
         }
 
-        let offset = 16 + (sig_size + 104) * size;
+        let offset = 16 + sig_reg_size * size;
         let batch_proof = BatchPath::from_bytes(&bytes[offset..])?;
 
         Ok(StmAggrSig {
-            signatures,
+            signatures: sig_reg_list,
             batch_proof,
         })
     }
 }
 
+impl CoreVerifier {
+    /// Setup a core verifier for given list of signers.
+    pub fn setup(public_signers: &[(VerificationKey, Stake)]) -> Self {
+        let mut total_stake: Stake = 0;
+        let mut unique_parties = HashSet::new();
+        for signer in public_signers.iter() {
+            let (res, overflow) = total_stake.overflowing_add(signer.1);
+            if overflow {
+                panic!("Total stake overflow");
+            }
+            total_stake = res;
+            unique_parties.insert(MTLeaf(signer.0, signer.1));
+        }
+
+        let eligible_parties: Vec<_> = unique_parties.into_iter().collect();
+        CoreVerifier {
+            eligible_parties,
+            total_stake,
+        }
+    }
+
+    fn preliminary_verify(
+        total_stake: &Stake,
+        signatures: &[StmSigRegParty],
+        parameters: &StmParameters,
+        msg: &[u8],
+    ) -> Result<(), CoreVerifierError> {
+        let mut nr_indices = 0;
+        let mut unique_indices = HashSet::new();
+
+        for sig_reg in signatures {
+            sig_reg
+                .sig
+                .check_indices(parameters, &sig_reg.reg_party.1, msg, total_stake)?;
+            for &index in &sig_reg.sig.indexes {
+                unique_indices.insert(index);
+                nr_indices += 1;
+            }
+        }
+
+        if nr_indices != unique_indices.len() {
+            return Err(CoreVerifierError::IndexNotUnique);
+        }
+        if (nr_indices as u64) < parameters.k {
+            return Err(CoreVerifierError::NoQuorum(nr_indices as u64, parameters.k));
+        }
+
+        Ok(())
+    }
+
+    fn map_sig_party(parties: &[RegParty], signatures: &[StmSig]) -> Vec<StmSigRegParty> {
```

Comment:
Let's add documentation even for private functions. Will make it easier for our future selfs. Also, you can extend a bit the todo. I forgot again what we are supposed to check 😓